src/core/Akka/Actor/ActorPath.cs

```diff
@@ -185,35 +192,72 @@ protected ActorPath(ActorPath parentPath, string name, long uid)
             Address = parentPath.Address;
             Uid = uid;
             Name = name;
+            _elements = new FastLazy<ActorPath, IList<string>>(FillElementsFunc, this);
         }
 
         /// <summary>
         /// Gets the uid.
         /// </summary>
         /// <value> The uid. </value>
         public long Uid { get; }
+        
+        private readonly FastLazy<ActorPath, IList<string>> _elements;
 
-        /// <summary>
-        /// Gets the elements.
-        /// </summary>
-        /// <value> The elements. </value>
-        public IReadOnlyList<string> Elements
+        private static readonly string[] _emptyElements = { };
+        private static readonly string[] _systemElements = { "system" };
+        private static readonly string[] _userElements = { "user" };
+
+        private static IList<string> FillElements(ActorPath actorPath)
         {
-            get
+            // fast path next three `if`
+            if(actorPath is RootActorPath)
+                return _emptyElements;
+            if (actorPath.Parent is RootActorPath)
             {
-                var current = this;
-                var elements = new List<string>();
-                while (!(current is RootActorPath))
+                if (actorPath.Name.Equals("system", StringComparison.Ordinal))
+                    return _systemElements;
+                if (actorPath.Name.Equals("user", StringComparison.Ordinal))
+                    return _userElements;
+                return new [] {actorPath.Name};
+            }
+            if (actorPath.Parent._elements.IsValueCreated)
+            {
+                var parentElems = actorPath.Parent._elements.Value;
+                var myElems = new string[parentElems.Count + 1];
+                parentElems.CopyTo(myElems, 0);
+                myElems[myElems.Length - 1] = actorPath.Name;
+                return myElems;
+            }
+
+            var current = actorPath;
+            var elements = new List<string>();
+            while (!(current is RootActorPath))
+            {
+                if (current._elements.IsValueCreated)
```

Comment:
`current._elements.IsValueCreated` will be `true` when someone already has called `ActorPath.Elements`. So we can stop iterating here and reuse `Parent`'s array. If it's not created yet dont touch it (preventing additional allocations).